File: c++/movezeros/move_zeros.cpp
#include <vector>
#include <iostream>
#include <algorithm>

using namespace std;

class Solution
{
public:
    vector<int> moveZeroes(vector<int> &nums)
    {
        int l = 0;
        int r = 0;
        for (r; r < nums.size(); r++)
        {
            if (nums[r])
            {
                std::swap(nums[r], nums[l]);
                l += 1;
            }
        }
        return nums;
    }
};

int main()
{
    Solution trial;
    vector<int> nums = {0, 1, 0, 3, 12};
    vector<int> result = trial.moveZeroes(nums);
    if (!result.empty())
    {
        cout << "[";
        for (size_t i = 0; i < result.size(); i++)
        {
            cout << result[i];
            if (i != result.size() - 1)
            {
                cout << ",";
            }
        }
        cout << "]" << endl;
    }
    else
    {
        cout << "Empty list" << endl;
    }
    return 0;
}

File: java/movezeros/move_zeros.java
package movezeros;

class Solution {
    public void moveZeroes(int[] nums) {
        int left = 0;
        for (int right = 0; right < nums.length; right++) {
            if (nums[right] != 0) {
                nums[left] = nums[right];
                left++;
            }
        }

        for (; left < nums.length; left++) {
            nums[left] = 0;
        }
    }
}